File: bw_convert.py
#!/usr/bin/env python3
"""Convert a PNG to black-and-white using a darkness threshold."""
import argparse
from pathlib import Path

import numpy as np
from PIL import Image, ImageFilter

DESPECKLE_SIZE = 3  # odd integer >= 3; controls default median filter kernel


def main():
    parser = argparse.ArgumentParser(
        description=(
            "Convert a PNG to black and white based on a darkness threshold. "
            "Outputs three variants: plain BW, BW with white transparent, "
            "and inverted BW with black transparent (white only on clear)."
        )
    )
    parser.add_argument("input", help="Input PNG file path")
    parser.add_argument(
        "threshold",
        type=float,
        help=(
            "Darkness threshold as a percentage (0–100). "
            "Pixels with darkness above this become black, the rest white."
        ),
    )
    parser.add_argument(
        "-o", "--output-dir",
        help="Directory for output files (default: same directory as input)",
    )
    parser.add_argument(
        "--no-despeckle",
        action="store_true",
        help="Disable the default median-filter despeckling step.",
    )
    args = parser.parse_args()

    if not 0.0 <= args.threshold <= 100.0:
        parser.error("Threshold must be between 0 and 100.")

    input_path = Path(args.input)
    if not input_path.exists():
        parser.error(f"File not found: {args.input}")

    out_dir = Path(args.output_dir) if args.output_dir else input_path.parent
    stem = input_path.stem

    # Load image; composite transparent inputs onto white before grayscaling
    img = Image.open(input_path)
    if img.mode == "RGBA":
        bg = Image.new("RGB", img.size, (255, 255, 255))
        bg.paste(img, mask=img.split()[3])
        gray = bg.convert("L")
    else:
        gray = img.convert("L")

    cutoff = 255.0 * (1.0 - args.threshold / 100.0)
    arr = np.array(gray, dtype=np.float32)
    bw = Image.fromarray(
        np.where(arr < cutoff, np.uint8(0), np.uint8(255)).astype(np.uint8), "L"
    )

    if not args.no_despeckle:
        bw = bw.filter(ImageFilter.MedianFilter(size=DESPECKLE_SIZE))

    # 1. Plain BW
    path_bw = out_dir / f"{stem}_bw.png"
    bw.convert("RGB").save(path_bw, format="PNG")
    print(f"Saved: {path_bw}")

    # 2. BW with white transparent (black visible on clear)
    path_transparent = out_dir / f"{stem}_bw_transparent.png"
    a = np.array(bw.convert("RGBA"))
    a[a[:, :, 0] == 255, 3] = 0
    Image.fromarray(a, "RGBA").save(path_transparent, format="PNG")
    print(f"Saved: {path_transparent}")

    # 3. Inverted BW with black transparent (white visible on clear)
    path_inverted = out_dir / f"{stem}_bw_inverted.png"
    inverted = Image.fromarray(255 - np.array(bw), "L")
    a_inv = np.array(inverted.convert("RGBA"))
    a_inv[a_inv[:, :, 0] == 0, 3] = 0
    Image.fromarray(a_inv, "RGBA").save(path_inverted, format="PNG")
    print(f"Saved: {path_inverted}")


if __name__ == "__main__":
    main()

File: web/src/lib/bwConvert.js
import { compositeOnWhite, imageDataToDataURL } from './imageUtils.js'

/**
 * Apply a 3×3 median filter to a grayscale Uint8ClampedArray.
 * Matches PIL.ImageFilter.MedianFilter(size=3).
 */
function medianFilter3(gray, width, height) {
  const out = new Uint8ClampedArray(gray.length)
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const neighbors = []
      for (let dy = -1; dy <= 1; dy++) {
        for (let dx = -1; dx <= 1; dx++) {
          const ny = Math.min(height - 1, Math.max(0, y + dy))
          const nx = Math.min(width - 1, Math.max(0, x + dx))
          neighbors.push(gray[ny * width + nx])
        }
      }
      neighbors.sort((a, b) => a - b)
      out[y * width + x] = neighbors[4] // median of 9
    }
  }
  return out
}

/**
 * Convert an ImageData to BW using a darkness threshold (0–100).
 * Returns { plain, transparent, inverted } as data URLs.
 * Matches bw_convert.py behavior exactly.
 */
export function bwConvert(imageData, threshold, despeckle = true) {
  const { width, height } = imageData

  // Composite onto white if needed
  const rgb = compositeOnWhite(imageData)
  const src = rgb.data

  // Grayscale
  const gray = new Uint8ClampedArray(width * height)
  for (let i = 0; i < width * height; i++) {
    const o = i * 4
    gray[i] = Math.round(0.299 * src[o] + 0.587 * src[o + 1] + 0.114 * src[o + 2])
  }

  // Threshold: cutoff = 255 * (1 - threshold/100)
  const cutoff = 255 * (1 - threshold / 100)
  let bw = new Uint8ClampedArray(width * height)
  for (let i = 0; i < gray.length; i++) {
    bw[i] = gray[i] < cutoff ? 0 : 255
  }

  if (despeckle) {
    bw = medianFilter3(bw, width, height)
  }

  // 1. Plain BW (RGB)
  const plainData = new Uint8ClampedArray(width * height * 4)
  for (let i = 0; i < width * height; i++) {
    const o = i * 4
    plainData[o] = plainData[o + 1] = plainData[o + 2] = bw[i]
    plainData[o + 3] = 255
  }
  const plain = imageDataToDataURL(new ImageData(plainData, width, height))

  // 2. BW with white → transparent (black visible on clear)
  const transData = new Uint8ClampedArray(width * height * 4)
  for (let i = 0; i < width * height; i++) {
    const o = i * 4
    transData[o] = transData[o + 1] = transData[o + 2] = bw[i]
    transData[o + 3] = bw[i] === 255 ? 0 : 255
  }
  const transparent = imageDataToDataURL(new ImageData(transData, width, height))

  // 3. Inverted BW with black → transparent (white visible on clear)
  const invData = new Uint8ClampedArray(width * height * 4)
  for (let i = 0; i < width * height; i++) {
    const o = i * 4
    const v = 255 - bw[i]
    invData[o] = invData[o + 1] = invData[o + 2] = v
    invData[o + 3] = v === 0 ? 0 : 255
  }
  const inverted = imageDataToDataURL(new ImageData(invData, width, height))

  return { plain, transparent, inverted }
}
